composer/callbacks/oom_observer.py

```diff
@@ -0,0 +1,177 @@
+
+            >>> from composer import Trainer
+            >>> from composer.callbacks import OOMObserver
+            >>> # constructing trainer object with this callback
+            >>> trainer = Trainer(
+            ...     model=model,
+            ...     train_dataloader=train_dataloader,
+            ...     eval_dataloader=eval_dataloader,
+            ...     optimizers=optimizer,
+            ...     max_duration="1ep",
+            ...     callbacks=[OOMObserver()],
+            ... )
+
+    .. note::
+        OOMObserver is only supported for GPU devices.
+
+    Args:
+        max_entries (int, optional): Maximum number of memory alloc/free events to record. Defaults to 100000.
+        folder (str, optional): A format string describing the folder containing the memory visualization files.
+            Defaults to ``'{{run_name}}/torch_traces'``.
+        filename (str, optional): A format string describing the prefix used to name the memory visualization files.
+            Defaults to ``'rank{{rank}}_oom'``.
+        remote_file_name (str, optional): A format string describing the prefix for the memory visualization remote file name.
+            Defaults to ``'{{run_name}}/oom_traces/rank{{rank}}_oom'``.
+
+            Whenever a trace file is saved, it is also uploaded as a file according to this format string.
+            The same format variables as for ``filename`` are available.
+
+            .. seealso:: :doc:`Uploading Files</trainer/file_uploading>` for notes for file uploading.
+
+            Leading slashes (``'/'``) will be stripped.
+
+            To disable uploading trace files, set this parameter to ``None``.
+        overwrite (bool, optional): Whether to override existing memory snapshots. Defaults to False.
+
+            If False, then the trace folder as determined by ``folder`` must be empty.
+    """
+
+    def __init__(
+        self,
+        max_entries: int = 100000,
+        folder: str = '{run_name}/torch_traces',
+        filename: str = 'rank{rank}_oom',
+        remote_file_name: Optional[str] = '{run_name}/oom_traces/rank{rank}_oom',
+        overwrite: bool = False,
+    ) -> None:
+        self.max_entries = max_entries
+        self.folder = folder
+        self.folder_name = None
+        self.filename = filename
+        self.remote_file_name = remote_file_name
+        self.overwrite = overwrite
+        if remote_file_name:
+            self.remote_file_name = remote_file_name
+            _, _, self.remote_path_in_bucket = parse_uri(remote_file_name)
+        else:
+            self.remote_path_in_bucket = None
+
+        if version.parse(torch.__version__.split('.dev')[0]) >= version.parse('2.1.0'):  # type: ignore
+            # OOMObserver  is only supported in torch v2.1.0-rc1 or higher
+            self._enabled = True
+        else:
+            self._enabled = False
+            log.warning('OOMObserver is supported after PyTorch 2.1.0. Skipping oom observer callback.')
```

Comment:
```suggestion
            warnings.warn('OOMObserver is supported after PyTorch 2.1.0. Disabling OOMObserver callback.')
```
I think this should be warnings.warn given it may require user action.

Reply:
warnings.warn will break cpu callback tests where user warnings are not catched. If training without oom observer support due to older pt version, the user shall just get notified in the log, no action shall be expected.

Reply:
I believe the CPU tests let you attach marks to filter these warnings?

Reply:
I don't have a super strong opinion here. But, I do think we should be consistent, and elsewhere we seem to prefer warnings.warn?

Reply:
warnings.warn is for when the user can do something, log.warning is when they can't. should use `pytest.mark.filterwarnings` to ignore

Reply:
see https://docs.python.org/3/howto/logging.html

Reply:
do we expect users to take any actions? I feel pt version is not something users will switch just because they saw oom observer is not supported. More like they use some pt version and are told oom observer is not supported and disabled.

Reply:
I think this is an actionable task. It is "avoidable" (from the logging docs)

Reply:
Yeah, this includes anything that the user _could_ do something about, not taking into account the likelihood of them doing that (at least that is my interpretation). `log.warning` is when the library just wants to tell you something important, but youcan't do anything about it.

Reply:
sounds good. will make the change